docs/Experimental Features.md

```diff
@@ -0,0 +1,217 @@
+
+```elixir
+# Customize non-Elixir parts of the firmware. See
+# https://hexdocs.pm/nerves/advanced-configuration.html for details.
+
+config :nerves, :firmware,
+  rootfs_overlay: "rootfs_overlay",
+  mksquashfs_flags: ["-noI", "-noId", "-noD", "-noF", "-noX"]
+```
+
+Patch sizes can also be optimized by configuring the build system's
+`source_date_epoch` date. This will help with reproducible builds by preventing
+timestamps modifications from affecting the output bit representation.
+
+```elixir
+# Set the SOURCE_DATE_EPOCH date for reproducible builds.
+# See https://reproducible-builds.org/docs/source-date-epoch/ for more information
+
+config :nerves, source_date_epoch: "1596027629"
+```
+
+### Testing firmware patches locally
+
+Create a new project using `mix nerves.new <project name>` and apply the steps
+listed in the `Preparing your project` section. Then, choose a target, in this
+example, I will be using a Raspberry Pi Zero W `rpi0` and building an app
+called `test_patch`.
+
+```
+export MIX_TARGET=rpi0
+mix deps.get
+```
+
+Create your initial firmware and burn it to an SD card
+
+```
+mix firmware.burn
+```
+
+Connect the SD card and power on the device by connecting a micro USB cable to
+the host USB port on the Raspberry Pi. You can ssh into the device at
+`nerves.local` and you should get an IEX prompt.
+
+```
+ssh nerves.local
+
+Interactive Elixir (1.10.3) - press Ctrl+C to exit (type h() ENTER for help)
+Toolshed imported. Run h(Toolshed) for more info.
+RingLogger is collecting log messages from Elixir and Linux. To see the
+messages, either attach the current IEx session to the logger:
+
+  RingLogger.attach
+
+or print the next messages in the log:
+
+  RingLogger.next
+
+iex(1)> TestPatch.hello
+:world
+```
+
+Make some changes to the function. Open `lib/<app_name>.ex` and modify the
+`hello/0` function.
+
+```elixir
+def hello do
+  :patched
+end
+```
+
+Now lets generate a patch firmware.
+
+`mix firmware.patch`
+
+You should see output similar to the following:
+
+```
+Finished generating patch firmware
+
+Source
+test_patch/_build/rpi0_dev/nerves/images/test_patch.fw
+uuid: 6cf7f75f-eb93-5a91-e28c-fd414602b6e7"
+
+size: 32053029 bytes
+
+Target
+nerves-project/tests/test_patch/_build/rpi0_dev/nerves/images/patch/target.fw
+uuid: 69752f24-291f-5f00-4ad3-ca359017009f"
+
+size: 32053263 bytes
+
+Patch
+test_patch/_build/rpi0_dev/nerves/images/patch.fw
+size: 4421811 bytes
+```
+
+Lets update the device using the patch file.
+
+```
+mix upload --firmware /path/to/test_patch/_build/rpi0_dev/nerves/images/patch.fw
+```
+
+The size difference between the `Target` output firmware size `32053263` and the
+patched firmware size `4421811` has a pretty significant size reduction. For
+such a small change, we might expect more. A lot of this size come from the
+files that are also included in the firmware that are not currently being patched
+such as the linux kernel and other files that do not change frequently.
```

Comment:
I think this is a true statement especially on the Raspberry Pi where there are a lot of bytes trapped in the boot partition. Can this be tried with `MIX_ENV=prod`, though? It doesn't seem like anyone would be pushing dev builds to remote devices. Of course, I do development with `MIX_ENV=prod` since firmware updates are faster and I don't use `h/1` much on device.

Reply:
I updated this with `prod` env stats.